File: Ejercicios/Ejercicio3/3.py
# Las subclases deben ser intercambiables por sus clases base sin alterar el comportamiento del programa
class Forma:
    def calcular_area(self):
        pass

class Rectangulo(Forma):
    def __init__(self, ancho, altura):
        self.ancho = ancho
        self.altura = altura

    def calcular_area(self):
        return self.ancho * self.altura

class Cuadrado(Rectangulo):
    def __init__(self, lado):
        super().__init__(lado, lado)

def calculararea(forma):
    return forma.calculararea()


rectangulo = Rectangulo(3, 4)
cuadrado = Cuadrado(5)

print("Area del rectangulo:", calculararea(rectangulo))
print("Area del cuadrado:", calculararea(cuadrado))

File: Ejercicios/Ejercicio3/3.cpp
#include <iostream> //las subclases deben ser intercambiables por sus clases base sin alterar el comportamiento del programa
using namespace std;

class Forma 
{
public:
    virtual double calcularArea() const = 0;
};


class Rectangulo : public Forma 
{
private:
    double ancho;
    double altura;

public:
    Rectangulo(double ancho, double altura) : ancho(ancho), altura(altura) {}

    double calcularArea() const override 
    {
        return ancho * altura;
    }
};

class Cuadrado : public Rectangulo 
{
public:
    Cuadrado(double lado) : Rectangulo(lado, lado) {}
};


double calcularArea(const Forma& forma) 
{
    return forma.calcularArea();
}

int main() 
{
    Rectangulo rectangulo(3, 4);
    Cuadrado cuadrado(5);

    cout << "Area del rectangulo: " << calcularArea(rectangulo) << endl;
    cout << "Area del cuadrado: " << calcularArea(cuadrado) << endl;

    return 0;
}
